src/serde/ser.rs

```diff
@@ -0,0 +1,621 @@
+where
+    T: Serialize,
+{
+    let mut cursor = Cursor::new(Vec::new());
+    let mut serializer = Serializer {
+        writer: TextWriterBuilder::new(TextKind::Compact).build(&mut cursor)?,
+    };
+
+    value.serialize(&mut serializer)?;
+    serializer.writer.flush()?;
+    drop(serializer);
+
+    let bytes = cursor.get_ref().clone();
+
+    match String::from_utf8(bytes) {
+        Ok(data) => Ok(data),
+        Err(e) => IonResult::encoding_error(e.to_string()),
+    }
+}
+
+/// Serialize an object into Ion binary format
+pub fn to_binary<T>(value: &T) -> IonResult<Vec<u8>>
+where
+    T: Serialize,
+{
+    let mut cursor = Cursor::new(Vec::new());
+    let mut serializer = Serializer {
+        writer: BinaryWriterBuilder::new().build(&mut cursor)?,
+    };
+
+    value.serialize(&mut serializer)?;
+    serializer.writer.flush()?;
+    drop(serializer);
+
+    Ok(cursor.get_ref().clone())
+}
+
+/// Implements a standard serializer for Ion
+pub struct Serializer<E> {
+    pub(crate) writer: E,
+}
+
+impl<'a, E> ser::Serializer for &'a mut Serializer<E>
+where
+    E: IonWriter,
+{
+    type Ok = ();
+    type Error = IonError;
+
+    type SerializeSeq = Self;
+    type SerializeTuple = Self;
+    type SerializeTupleStruct = Self;
+    type SerializeTupleVariant = Self;
+    type SerializeMap = Self;
+    type SerializeStruct = Self;
+    type SerializeStructVariant = Self;
+
+    /// Serialize a boolean to a bool value
+    fn serialize_bool(self, v: bool) -> Result<Self::Ok, Self::Error> {
+        self.writer.write_bool(v)
+    }
+
+    /// Serialize all integer types using the `Integer` intermediary type.
+    fn serialize_i8(self, v: i8) -> Result<Self::Ok, Self::Error> {
+        self.writer.write_int(&Int::from(v))
+    }
+
+    /// Serialize all integer types using the `Integer` intermediary type.
+    fn serialize_u8(self, v: u8) -> Result<Self::Ok, Self::Error> {
+        self.writer.write_int(&Int::from(v))
+    }
+
+    /// Serialize all integer types using the `Integer` intermediary type.
+    fn serialize_i16(self, v: i16) -> Result<Self::Ok, Self::Error> {
+        self.writer.write_int(&Int::from(v))
+    }
+
+    /// Serialize all integer types using the `Integer` intermediary type.
+    fn serialize_u16(self, v: u16) -> Result<Self::Ok, Self::Error> {
+        self.writer.write_int(&Int::from(v))
+    }
+
+    /// Serialize all integer types using the `Integer` intermediary type.
+    fn serialize_i32(self, v: i32) -> Result<Self::Ok, Self::Error> {
+        self.writer.write_int(&Int::from(v))
+    }
+
+    /// Serialize all integer types using the `Integer` intermediary type.
+    fn serialize_u32(self, v: u32) -> Result<Self::Ok, Self::Error> {
+        self.writer.write_int(&Int::from(v))
+    }
+
+    /// Serialize all integer types using the `Integer` intermediary type.
+    fn serialize_i64(self, v: i64) -> Result<Self::Ok, Self::Error> {
+        self.writer.write_int(&Int::from(v))
```

Comment:
For all of these that fit in an `i64` (which is everything but `u64`) you can use `self.writer.write_i64()` to avoid making an intermediate `Int`.